src/wp-includes/blocks.php

```diff
@@ -853,18 +853,24 @@ function get_comment_delimited_block_content( $block_name, $block_attributes, $b
  *
  * @param array         $block    A representative array of a single parsed block object. See WP_Block_Parser_Block.
  * @param callable|null $callback Optional. Callback to run on each block in the tree before serialization. Default null.
+ *                                It is called with the following arguments: $block, $parent_block, $block_index, $chunk_index.
  * @return string String of rendered HTML.
  */
 function serialize_block( $block, $callback = null ) {
-	if ( is_callable( $callback ) ) {
-		$block = call_user_func( $callback, $block );
-	}
-
 	$block_content = '';
 
-	$index = 0;
-	foreach ( $block['innerContent'] as $chunk ) {
-		$block_content .= is_string( $chunk ) ? $chunk : serialize_block( $block['innerBlocks'][ $index++ ], $callback );
+	$block_index = 0;
+	foreach ( $block['innerContent'] as $chunk_index => $chunk ) {
+		if ( is_string( $chunk ) ) {
+			$block_content .= $chunk;
+		} else {
+			$inner_block = $block['innerBlocks'][ $block_index ];
+			if ( is_callable( $callback ) ) {
```

Comment:
I was a fan of your suggestion to split the loop, but without that, we could at least examine the impact of trapping `$is_callable = is_callable( $callback )` outside of the loop. it appears like `is_callable()` is a function call and not a PHP operand, so it could be notable in its cost, particularly on long posts.

in any case, here it seems a little odd to embed that check since `$callback` shouldn't change between iterations and there's no need to recompute it every time.